packages/valory/skills/abstract_round_abci/tests/test_common.py

```diff
@@ -51,3 +63,367 @@ def test_random_selection() -> None:
+        (
+            (
+                frozenset((f"test_p{i}" for i in range(4))),
+                frozenset(),
+                "test_p0",
+                "test_p1",
+            ),
+            (
+                frozenset((f"test_p{i}" for i in range(4))),
+                frozenset(),
+                "test_p1",
+                "test_p2",
+            ),
+            (
+                frozenset((f"test_p{i}" for i in range(4))),
+                frozenset(),
+                "test_p2",
+                "test_p3",
+            ),
+            (
+                frozenset((f"test_p{i}" for i in range(4))),
+                frozenset(),
+                "test_p3",
+                "test_p0",
+            ),
+            (
+                frozenset((f"test_p{i}" for i in range(4))),
+                frozenset((f"test_p{i}" for i in range(1))),
+                "test_p1",
+                "test_p2",
+            ),
+            (
+                frozenset((f"test_p{i}" for i in range(4))),
+                frozenset((f"test_p{i}" for i in range(4))),
+                "",
+                "",
+            ),
+        ),
+    )
+    def test_select_keeper(
+        self,
+        participants: FrozenSet[str],
+        blacklisted_keepers: FrozenSet[str],
+        most_voted_keeper_address: str,  # pylint: disable=unused-argument
+        expected_keeper: str,
+    ) -> None:
+        """Test `_select_keeper`."""
+        for sync_data_name in (
+            "participants",
+            "blacklisted_keepers",
+            "most_voted_keeper_address",
+        ):
+            setattr(
+                self.select_keeper_behaviour.context.state.synchronized_data,
+                sync_data_name,
+                locals()[sync_data_name],
+            )
+
+        select_keeper_method = (
+            self.select_keeper_behaviour._select_keeper  # pylint: disable=protected-access
+        )
+
+        if not participants - blacklisted_keepers:
+            with pytest.raises(
+                RuntimeError,
+                match="Cannot continue if all the keepers have been blacklisted!",
+            ):
+                select_keeper_method()
+            return
+
+        actual_keeper = select_keeper_method()
+        assert actual_keeper == expected_keeper
+
+    def test_async_act(self) -> None:
+        """Test `async_act`."""
+        gen = self.select_keeper_behaviour.async_act()
+
+        with mock.patch.object(
+            self.select_keeper_behaviour,
+            "_select_keeper",
+            return_value="test_keeper",
+        ), mock.patch.object(
+            self.select_keeper_behaviour,
+            "send_a2a_transaction",
+        ) as send_a2a_transaction_mocked, mock.patch.object(
+            self.select_keeper_behaviour,
+            "wait_until_round_end",
+        ) as wait_until_round_end_mocked, mock.patch.object(
+            self.select_keeper_behaviour,
+            "set_done",
+        ) as set_done_mocked:
```

Comment:
would be really great to have an extension of `as_context` that allows for this to become readable. 
```
with as_context(
    mocked_first_item,
    mocked_second_item,
    ) as mock:
    mock.mocked_first_item.assert_called_once()
    mock.mocked_second_item.assert_not_called()
    ...
```